soroban-sdk-macros/src/derive_enum.rs

```diff
@@ -67,75 +58,34 @@ pub fn derive_type_enum(
                 #discriminant_const_sym
                 #discriminant_const_u64
             };
-            if let Some(f) = field {
-                let spec_case = ScSpecUdtUnionCaseV0::TupleV0(
-                    ScSpecUdtUnionCaseTupleV0 {
-                    doc: docs_from_attrs(&v.attrs).try_into().unwrap(), // TODO: Truncate docs, or display friendly compile error.
-                    name: name.try_into().unwrap_or_else(|_| StringM::default()),
-                    type_: vec![
-                        match map_type(&f.ty) {
-                            Ok(t) => t,
-                            Err(e) => {
-                                errors.push(e);
-                                ScSpecTypeDef::I32
-                            }
-                        },
-                        ].try_into().unwrap()
-                    }
+            let has_fields = v.fields.iter().next().is_some();
+            if has_fields {
```

Comment:
Something I notice is that because we're continuing to use the number of fields to determine if the type is a unit vs tuple variant, we're not distinguishing between an empty tuple and a unit variant.

```rust
enum E {
    A,
    B(),
}
```

Because there's no distinction between the two, code generated versions of the contract will diverge and become:
```rust
enum E {
    A,
    B,
}
```

We could address the divergence by determining if we should map to the void or tuple union case on whether `v.fields` is a `Fields::Unnamed` or a `Fields::Unit` instead of on the number of fields. (https://docs.rs/syn/0.12.15/syn/enum.Fields.html)

But this is also such an edge case, that arguably we should just treat them both the same. There's really no benefit in encoding `B()` differently, and I doubt any other programming language will render the difference in a meaningful way. To many Rust developers the fact that `A` and `B()` are different is also probably a surprise. Also, if we encode `B()` as a tuple variant it is less efficient in the spec and in the conversions, so there's no win there.

So I think this is good as is, keeping to using `v.fields` length.

Thoughts anyone else?

Reply:
I forgot that empty tuple variants were possible - I would never think to write them!

That they won't round trip when re-rendered as rust seems like it could become a surprising problem in ways I can't imagine yet.

We could hedge and make empty-tuple variants an error for now, figure it out later. Hopefully nobody is using them already.

Reply:
It shocked me when empty tuple variants weren't the same as unit variants. I think there's a variety of ways they are a surprise and a foot gun. So doing as you suggest and erroring for now feels like a good plan, possibly even long term.

Reply:
Opened an issue for adding an error: https://github.com/stellar/rs-soroban-sdk/issues/861.